File: android/app/src/main/kotlin/com/homebase/android/ui/abwesenheit/Holidays.kt
package com.homebase.android.ui.abwesenheit

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.Month
import java.time.YearMonth
import java.time.format.TextStyle
import java.util.Locale

/**
 * German public-holiday computation per Bundesland + calendar date utils.
 * Ported verbatim from the design handoff (holidays.jsx / web holidays.ts) so the
 * mobile planner derives the exact same dates. Holidays are computed, never stored.
 */
object AbwCal {

    fun pad(n: Int): String = n.toString().padStart(2, '0')

    /** "YYYY-MM-DD" key for a date. */
    fun ymd(d: LocalDate): String = "${d.year}-${pad(d.monthValue)}-${pad(d.dayOfMonth)}"

    fun parse(s: String): LocalDate = LocalDate.parse(s)

    fun addDays(d: LocalDate, n: Int): LocalDate = d.plusDays(n.toLong())

    fun daysInMonth(year: Int, month0: Int): Int = YearMonth.of(year, month0 + 1).lengthOfMonth()

    /** ISO weekday: Mon = 1 … Sun = 7. */
    fun isoDow(d: LocalDate): Int = d.dayOfWeek.value

    fun isWeekend(d: LocalDate): Boolean =
        d.dayOfWeek == DayOfWeek.SATURDAY || d.dayOfWeek == DayOfWeek.SUNDAY

    /** Anonymous Gregorian Easter algorithm → Easter Sunday. */
    fun easter(year: Int): LocalDate {
        val a = year % 19
        val b = year / 100
        val c = year % 100
        val d = b / 4
        val e = b % 4
        val f = (b + 8) / 25
        val g = (b - f + 1) / 3
        val h = (19 * a + b - d - g + 15) % 30
        val i = c / 4
        val k = c % 4
        val l = (32 + 2 * e + 2 * i - h - k) % 7
        val m = (a + 11 * h + 22 * l) / 451
        val month = (h + l - 7 * m + 114) / 31 // 3 = March, 4 = April
        val day = ((h + l - 7 * m + 114) % 31) + 1
        return LocalDate.of(year, month, day)
    }

    /** Buß- und Bettag — the Wednesday before Nov 23. */
    private fun bussBettag(year: Int): LocalDate {
        var d = LocalDate.of(year, 11, 22)
        while (d.dayOfWeek != DayOfWeek.WEDNESDAY) d = d.minusDays(1)
        return d
    }

    data class GermanState(val code: String, val name: String)

    val STATES: List<GermanState> = listOf(
        GermanState("BW", "Baden-Württemberg"),
        GermanState("BY", "Bayern"),
        GermanState("BE", "Berlin"),
        GermanState("BB", "Brandenburg"),
        GermanState("HB", "Bremen"),
        GermanState("HH", "Hamburg"),
        GermanState("HE", "Hessen"),
        GermanState("MV", "Mecklenburg-Vorpommern"),
        GermanState("NI", "Niedersachsen"),
        GermanState("NW", "Nordrhein-Westfalen"),
        GermanState("RP", "Rheinland-Pfalz"),
        GermanState("SL", "Saarland"),
        GermanState("SN", "Sachsen"),
        GermanState("ST", "Sachsen-Anhalt"),
        GermanState("SH", "Schleswig-Holstein"),
        GermanState("TH", "Thüringen"),
    )

    private val ALL: List<String> = STATES.map { it.code }

    fun stateName(code: String): String = STATES.find { it.code == code }?.name ?: code

    private val cache = HashMap<String, Map<String, String>>()

    // [en] is the standard, well-known English name where one exists; null keeps the German name in
    // English mode too (region-specific/obscure feast days with no clean English equivalent — #204).
    private data class HolidayDef(val date: LocalDate, val de: String, val en: String?, val states: List<String>)

    /**
     * → { "YYYY-MM-DD": "Feiertagsname", … } for a given year + Bundesland, in [locale]'s language.
     * Well-known holidays get their standard English name (Karfreitag→Good Friday, …); the rest stay
     * German in both languages (#204). The names are display-only — date lookups elsewhere key on the
     * date, not the name. Cached per year:state:lang; defaults to the active app locale.
     */
    fun holidays(year: Int, state: String, locale: Locale = Locale.getDefault()): Map<String, String> {
        val en = locale.language == Locale.ENGLISH.language
        val key = "$year:$state:${if (en) "en" else "de"}"
        cache[key]?.let { return it }
        val e = easter(year)
        fun off(n: Int) = e.plusDays(n.toLong())
        val defs = listOf(
            HolidayDef(LocalDate.of(year, 1, 1), "Neujahr", "New Year's Day", ALL),
            HolidayDef(LocalDate.of(year, 1, 6), "Heilige Drei Könige", "Epiphany", listOf("BW", "BY", "ST")),
            HolidayDef(LocalDate.of(year, 3, 8), "Internationaler Frauentag", "International Women's Day", listOf("BE", "MV")),
            HolidayDef(off(-2), "Karfreitag", "Good Friday", ALL),
            HolidayDef(off(0), "Ostersonntag", "Easter Sunday", listOf("BB")),
            HolidayDef(off(1), "Ostermontag", "Easter Monday", ALL),
            HolidayDef(LocalDate.of(year, 5, 1), "Tag der Arbeit", "Labour Day", ALL),
            HolidayDef(off(39), "Christi Himmelfahrt", "Ascension Day", ALL),
            HolidayDef(off(49), "Pfingstsonntag", "Whit Sunday", listOf("BB")),
            HolidayDef(off(50), "Pfingstmontag", "Whit Monday", ALL),
            HolidayDef(off(60), "Fronleichnam", "Corpus Christi", listOf("BW", "BY", "HE", "NW", "RP", "SL")),
            HolidayDef(LocalDate.of(year, 8, 15), "Mariä Himmelfahrt", "Assumption Day", listOf("SL", "BY")),
            // Thuringia-specific commemoration — no standard English name, keep German.
            HolidayDef(LocalDate.of(year, 9, 20), "Weltkindertag", null, listOf("TH")),
            HolidayDef(LocalDate.of(year, 10, 3), "Tag der Deutschen Einheit", "German Unity Day", ALL),
            HolidayDef(LocalDate.of(year, 10, 31), "Reformationstag", "Reformation Day", listOf("BB", "HB", "HH", "MV", "NI", "SN", "ST", "SH", "TH")),
            HolidayDef(LocalDate.of(year, 11, 1), "Allerheiligen", "All Saints' Day", listOf("BW", "BY", "NW", "RP", "SL")),
            // Saxony-specific Protestant day of repentance — no common English name, keep German.
            HolidayDef(bussBettag(year), "Buß- und Bettag", null, listOf("SN")),
            HolidayDef(LocalDate.of(year, 12, 25), "1. Weihnachtstag", "Christmas Day", ALL),
            HolidayDef(LocalDate.of(year, 12, 26), "2. Weihnachtstag", "Boxing Day", ALL),
        )
        val map = LinkedHashMap<String, String>()
        defs.forEach { def -> if (state in def.states) map[ymd(def.date)] = if (en) def.en ?: def.de else def.de }
        cache[key] = map
        return map
    }

    /** All date-strings of a year, in order. */
    fun yearDates(year: Int): List<String> {
        val out = ArrayList<String>(366)
        for (m in 0 until 12) {
            val n = daysInMonth(year, m)
            for (d in 1..n) out.add("$year-${pad(m + 1)}-${pad(d)}")
        }
        return out
    }

    // Month/weekday names follow the active app locale (#204): German base, English when switched.
    // The grid renders these by index, so they're exposed as 0-based / ISO-ordered lists built from
    // java.time display names for the given (default) locale.

    /** Full month names, January-first (index 0 = January), in [locale]. */
    fun monFull(locale: Locale = Locale.getDefault()): List<String> =
        Month.entries.map { it.getDisplayName(TextStyle.FULL, locale) }

    /** Abbreviated month names, January-first, in [locale]. */
    fun monAbbr(locale: Locale = Locale.getDefault()): List<String> =
        Month.entries.map { it.getDisplayName(TextStyle.SHORT, locale) }

    /** Short weekday names in ISO order (index 0 = Monday), in [locale]. */
    fun wdMin(locale: Locale = Locale.getDefault()): List<String> =
        DayOfWeek.entries.map { it.getDisplayName(TextStyle.SHORT, locale) }

    /** Full weekday names in ISO order (index 0 = Monday), in [locale]. */
    fun wdLong(locale: Locale = Locale.getDefault()): List<String> =
        DayOfWeek.entries.map { it.getDisplayName(TextStyle.FULL, locale) }

    /** "Do, 11. Juni 2026" / "Thu, 11 June 2026" — short-weekday sheet title for a date string. */
    fun dayTitle(ds: String, locale: Locale = Locale.getDefault()): String {
        val d = parse(ds)
        val wd = d.dayOfWeek.getDisplayName(TextStyle.SHORT, locale)
        val mon = d.month.getDisplayName(TextStyle.FULL, locale)
        val en = locale.language == Locale.ENGLISH.language
        return if (en) "$wd, ${d.dayOfMonth} $mon ${d.year}" else "$wd, ${d.dayOfMonth}. $mon ${d.year}"
    }

    /** "31.3." — day.month. for carry-over expiry chips. */
    fun ddmm(ds: String?): String {
        if (ds.isNullOrBlank()) return ""
        val d = parse(ds)
        return "${d.dayOfMonth}.${d.monthValue}."
    }
}

File: web/src/components/abwesenheit/holidays.ts
// German public-holiday computation + calendar date utils.
// Ported from the original design handoff (HBcal). Pure, framework-free.

export const pad = (n: number): string => String(n).padStart(2, '0')

export const ymd = (d: Date): string => `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`

/** Parse "YYYY-MM-DD" as a *local* date (noon, to dodge DST edges). */
export const parse = (s: string): Date => {
  const [y, m, d] = s.split('-').map(Number)
  return new Date(y, m - 1, d, 12, 0, 0, 0)
}

export const addDays = (d: Date, n: number): Date => {
  const r = new Date(d)
  r.setDate(r.getDate() + n)
  return r
}

export const daysInMonth = (year: number, month0: number): number => new Date(year, month0 + 1, 0).getDate()

/** ISO weekday: Mon = 1 … Sun = 7 */
export const isoDow = (d: Date): number => ((d.getDay() + 6) % 7) + 1

export const isWeekend = (d: Date): boolean => {
  const w = d.getDay()
  return w === 0 || w === 6
}

/** Anonymous Gregorian Easter algorithm → Date of Easter Sunday. */
export function easter(year: number): Date {
  const a = year % 19
  const b = Math.floor(year / 100)
  const c = year % 100
  const d = Math.floor(b / 4)
  const e = b % 4
  const f = Math.floor((b + 8) / 25)
  const g = Math.floor((b - f + 1) / 3)
  const h = (19 * a + b - d - g + 15) % 30
  const i = Math.floor(c / 4)
  const k = c % 4
  const l = (32 + 2 * e + 2 * i - h - k) % 7
  const m = Math.floor((a + 11 * h + 22 * l) / 451)
  const month = Math.floor((h + l - 7 * m + 114) / 31) // 3 = March, 4 = April
  const day = ((h + l - 7 * m + 114) % 31) + 1
  return new Date(year, month - 1, day, 12)
}

/** Buß- und Bettag — the Wednesday before Nov 23. */
function bussBettag(year: number): Date {
  const d = new Date(year, 10, 22, 12)
  while (d.getDay() !== 3) d.setDate(d.getDate() - 1)
  return d
}

export interface GermanState {
  code: string
  name: string
}

export const STATES: GermanState[] = [
  { code: 'BW', name: 'Baden-Württemberg' },
  { code: 'BY', name: 'Bayern' },
  { code: 'BE', name: 'Berlin' },
  { code: 'BB', name: 'Brandenburg' },
  { code: 'HB', name: 'Bremen' },
  { code: 'HH', name: 'Hamburg' },
  { code: 'HE', name: 'Hessen' },
  { code: 'MV', name: 'Mecklenburg-Vorpommern' },
  { code: 'NI', name: 'Niedersachsen' },
  { code: 'NW', name: 'Nordrhein-Westfalen' },
  { code: 'RP', name: 'Rheinland-Pfalz' },
  { code: 'SL', name: 'Saarland' },
  { code: 'SN', name: 'Sachsen' },
  { code: 'ST', name: 'Sachsen-Anhalt' },
  { code: 'SH', name: 'Schleswig-Holstein' },
  { code: 'TH', name: 'Thüringen' },
]

export const ALL: string[] = STATES.map((s) => s.code)
export const stateName = (code: string): string => STATES.find((s) => s.code === code)?.name ?? code

const _cache: Record<string, Record<string, string>> = {}

/** → { "YYYY-MM-DD": "Feiertagsname", … } for a given year + Bundesland. */
export function holidays(year: number, state: string): Record<string, string> {
  const key = `${year}:${state}`
  if (_cache[key]) return _cache[key]
  const E = easter(year)
  const off = (n: number) => addDays(E, n)
  const defs: { d: Date; name: string; s: string[] }[] = [
    { d: new Date(year, 0, 1, 12), name: 'Neujahr', s: ALL },
    { d: new Date(year, 0, 6, 12), name: 'Heilige Drei Könige', s: ['BW', 'BY', 'ST'] },
    { d: new Date(year, 2, 8, 12), name: 'Internationaler Frauentag', s: ['BE', 'MV'] },
    { d: off(-2), name: 'Karfreitag', s: ALL },
    { d: off(0), name: 'Ostersonntag', s: ['BB'] },
    { d: off(1), name: 'Ostermontag', s: ALL },
    { d: new Date(year, 4, 1, 12), name: 'Tag der Arbeit', s: ALL },
    { d: off(39), name: 'Christi Himmelfahrt', s: ALL },
    { d: off(49), name: 'Pfingstsonntag', s: ['BB'] },
    { d: off(50), name: 'Pfingstmontag', s: ALL },
    { d: off(60), name: 'Fronleichnam', s: ['BW', 'BY', 'HE', 'NW', 'RP', 'SL'] },
    { d: new Date(year, 7, 15, 12), name: 'Mariä Himmelfahrt', s: ['SL', 'BY'] },
    { d: new Date(year, 8, 20, 12), name: 'Weltkindertag', s: ['TH'] },
    { d: new Date(year, 9, 3, 12), name: 'Tag der Deutschen Einheit', s: ALL },
    { d: new Date(year, 9, 31, 12), name: 'Reformationstag', s: ['BB', 'HB', 'HH', 'MV', 'NI', 'SN', 'ST', 'SH', 'TH'] },
    { d: new Date(year, 10, 1, 12), name: 'Allerheiligen', s: ['BW', 'BY', 'NW', 'RP', 'SL'] },
    { d: bussBettag(year), name: 'Buß- und Bettag', s: ['SN'] },
    { d: new Date(year, 11, 25, 12), name: '1. Weihnachtstag', s: ALL },
    { d: new Date(year, 11, 26, 12), name: '2. Weihnachtstag', s: ALL },
  ]
  const map: Record<string, string> = {}
  defs.forEach((def) => {
    if (def.s.includes(state)) map[ymd(def.d)] = def.name
  })
  _cache[key] = map
  return map
}

/** All date-strings of a year. */
export function yearDates(year: number): string[] {
  const out: string[] = []
  for (let m = 0; m < 12; m++) {
    const n = daysInMonth(year, m)
    for (let d = 1; d <= n; d++) out.push(ymd(new Date(year, m, d, 12)))
  }
  return out
}

export const MON_FULL = ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember']
export const MON_ABBR = ['Jan', 'Feb', 'Mär', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez']
export const WD_MIN = ['Mo', 'Di', 'Mi', 'Do', 'Fr', 'Sa', 'So'] // ISO order
export const WD_LONG = ['Sonntag', 'Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag'] // JS getDay order
